### PR TITLE
Add cluster setting cluster.restrict.index.replication_type t…

### LINKED ISSUE #10868: Introduce a cluster setting that would help restrict replication type in a cluster
**Is your feature request related to a problem? Please describe.**
Currently, we have a cluster setting that sets the default replication type for any index in the cluster, namely - CLUSTER_SETTING_REPLICATION_TYPE = "cluster.indices.replication.strategy". This can be overridden by the index.replication.type setting currently. 

If we set it to false by default, working of cluster will be as usual and only enforce restriction on the index setting of replication type when it is set to true.

**Describe the solution you'd like**
Introducing a new cluster setting would help enforce a replication strategy that cannot be overridden by the index replication type.

### PR DESCRIPTION
**Pull request by Poojita-Raj**
…o restrict setting of index setting replication type

### Description
Introduces a cluster level setting CLUSTER_RESTRICT_INDEX_REPLICATION_TYPE_SETTING = cluster.restrict.index.replication_type that can restrict setting of index.replication.type setting in a cluster. 
Default value = false. 
Throws an exception when we try to set the index replication type when the cluster restrict setting is enabled. 

### Related Issues
Resolves #10868 

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationClusterSettingIT.java [hunk 1]
```diff
@@ -123,4 +124,31 @@ public void testIndexReplicationSettingOverridesDocRepClusterSetting() throws Ex
...
+        String expectedExceptionMsg =
+            "Validation Failed: 1: index setting [index.replication.type] is not allowed to be set as [cluster.restrict.index.replication_type=true];";
+        String clusterManagerName = internalCluster().startNode(
+            Settings.builder().put(CLUSTER_RESTRICT_INDEX_REPLICATION_TYPE_SETTING.getKey(), setRestrict).build()
+        );
+        internalCluster().startDataOnlyNodes(1);
+
+        // Test create index fails
+        Settings indexSettings = Settings.builder().put(indexSettings()).put(SETTING_REPLICATION_TYPE, replicationType).build();
+        try {
```
Comment (mch2) [inline server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationClusterSettingIT.java:145]: A nit - I'd prefer checking the value of setRestrict and then using expectThrows with the expected Exception type to catching any Exception:

```java
if (setRestrict) {
  expectThrows(..., () -> createIndex(INDEX_NAME, indexSettings);
} else {
  createIndex(INDEX_NAME, indexSettings);
}
```

File: server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java [hunk 2]
```diff
@@ -1252,14 +1252,16 @@ List<String> getIndexSettingsValidationErrors(
         if (forbidPrivateIndexSettings) {
             validationErrors.addAll(validatePrivateSettingsNotExplicitlySet(settings, indexScopedSettings));
         }
+        Optional<String> error = validateIndexReplicationTypeSettings(settings, clusterService.getClusterSettings());
+        error.ifPresent(validationErrors::add);
```
Comment (mch2) [inline server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java:1256]: Would prefer we do this without reassigning the error variable:
```java
validateIndexReplicationTypeSettings(settings, clusterService.getClusterSettings()).ifPresent(validationErrors:add);
```

File: server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java [hunk 3]
```diff
@@ -1306,6 +1308,24 @@ private static List<String> validateIndexCustomPath(Settings settings, @Nullable
         return validationErrors;
     }
 
+    /**
+     * Validates {@code index.replication.type} is not set if {@code cluster.restrict.index.replication_type} is set to true.
+     *
+     * @param requestSettings settings passed in during index create request
+     * @param clusterSettings cluster setting
+     */
+    static Optional<String> validateIndexReplicationTypeSettings(Settings requestSettings, ClusterSettings clusterSettings) {
```
Comment (mch2) [inline server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java:1317]: i don't see this called directly in tests - can we make it private?

File: server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java [hunk 4]
```diff
@@ -1306,6 +1308,24 @@ private static List<String> validateIndexCustomPath(Settings settings, @Nullable
...
+        if (requestSettings.hasValue(SETTING_REPLICATION_TYPE) == false
```
Comment (mch2) [inline server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java:1318]: nit - I find the and case to be much more readable.

```java
        if (requestSettings.hasValue(SETTING_REPLICATION_TYPE)
            && clusterSettings.get(IndicesService.CLUSTER_RESTRICT_INDEX_REPLICATION_TYPE_SETTING)) {
        return Optional.of(
            "index setting [index.replication.type] is not allowed to be set as ["
                + IndicesService.CLUSTER_RESTRICT_INDEX_REPLICATION_TYPE_SETTING.getKey()
                + "=true]"
        );
        }
        return Optional.empty();

```

**Comment by Poojita-Raj** [comment]
Thanks for the review @mch2! Addressed all the changes, please take another look

**Review by mch2 — APPROVED** [review]
lgtm

**Review by Rishikesh1159 — APPROVED** [review]

**Pull request merged**